[kbn] Subscription tracking

janmonschke opened the pull request:
## Summary

With the introduction of more features that are part of licenses, we're also adding more upsells to Kibana. These upsells advertise the feature, they explain which license is required in order to use said feature and they will link the client to the subscription page. Take the upsell for more insights in the alert flyout as an example:

<img width="1584" alt="Screenshot 2022-10-18 at 16 39 52" src="https://user-images.githubusercontent.com/68591/197629708-17978c8b-595e-4797-b80a-59c799896509.png">

Upsells come in all different shapes. Somtimes they're just links, sometimes full pages and sometimes interactive popups. They are also used across all solutions in Kibana.

There is currently no specific tracking in place for these types of elements yet. Meaning we don't know how many people interact with them, how many custerms see them and how well they perform in terms of conversions.

It is technically already possible to analyze clicks on these elements as part of the regular Kibana  click tracking but it would require setting up queries with lots of  `data-test-subj` and `url` filters for the right click events. Even if we wanted to set up tracking dashboards with that data, we would still not know how often upsells are seen which is necessary to calculate their click-through-rate. That rate can give an indicator if an upsell performs well or if we might want to improve it in the future.

For that reason, I'm proposing a dedicated set of tracking methods to capture `impressions`  and `clicks` for upsells. No conversion tracking as of yet, but I will get back to that later.

This PR introduces the `@kbn/subscription-tracking` package. It leverages the `@kbn/analytics-client` package to send dedicated subscription tracking events.

It comes with a set of React components that automatically track impressions and click events. Consumers of those components only need to specify a `subscription context` that gives more details on the type of feature that is advertised and the location of the upsell.

```typescript
import { SubscriptionLink } from '@kbn/subscription-tracking';
import type { SubscriptionContext } from '@kbn/subscription-tracking';

const subscriptionContext: SubscriptionContext = {
  feature: 'threat-intelligence',
  source: 'security__threat-intelligence',
};

export const Paywall = () => {
  return (
    <div>
      <SubscriptionLink subscriptionContext={subscriptionContext}>
        Upgrade to Platinum to get this feature
      </SubscriptionLink>
    </div>
  )
}
```

The example above uses a `SubscriptionLink` which is a wrapper of `EuiLink` . So it behaves just like a normal link. Alternatively, upsells can also use a `SubscriptionButton` or `SubscriptionButtonEmpty` which wrap `EuiButton` and `EuiButtonEmpty` respectively.

When the link is mounted, it will send off an `impression` event with the given `subscriptionContext`. That piece of metadata consists of an identifier of the advertised feature (in this case `threat-intelligence`) and the `source` of the impression (in this case the `threat-intelligence` page in the `security` solution). `source` follows the following format: `{solution-identifier}__location-identifier`.

There are no special rules for how to name these identifiers but it's good practise to make sure that `feature` has the same value for all upsells advertising the same feature (e.g. use enums for features to prevent spelling mistakes).

Upon interaction with the upsell link/button, a special `click` event is sent, which, again, contains the same subscription context.

If you want to use the `subscription-tracking` elements in your solution, you have to set up a `SubscriptionTrackingProvider` in your plugin setup and register the tracking events on startup.  This PR contains an example for this setup for the security plugin and some of its sub-plugins.

## Next steps
- There are currently no dedicated tracking dashboards for these events which I am planning to set up in the future.
- Since I only had a week's worth of time, I did not manage to add conversion tracking. The addition of those events might be a lot harder as well since the current license flow does not integrate seamlessly into Kibana
- All upsells currently link to the license management page which currently does not inform customers about our license and cloud offering. It seems to me like a weak link in the subscription funnel and it would be great to improve on that page.
- potential improvement: Send `impression` event when the element becomes visible in the viewport instead of when the element is mounted

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

janmonschke committed:
chore: initialize package

janmonschke committed:
integrate subscription tracking into Kibana

janmonschke committed:
register events on the client

janmonschke committed:
stop faking the upsell

janmonschke committed:
remove unused code

janmonschke committed:
switch to double underscore

`::` does not work well with KQL

janmonschke committed:
add `analyticsClient.isEventTypeRegistered`

janmonschke committed:
don't register events if they're already registered

janmonschke committed:
use cooldown period for impression events

janmonschke committed:
Add SubscriptionButtonEmpty

janmonschke committed:
add subscription tracking to security_solution

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

janmonschke committed:
import types

janmonschke committed:
fix tests

janmonschke committed:
remove unused dependencies

janmonschke committed:
remove unused prop

pgayvallet reviewed (commented):
Core-owned changes looks good to me but I'll defer to @afharo

janmonschke committed:
provide mock provider where necessary

janmonschke committed:
documentation

janmonschke committed:
fix tests

janmonschke committed:
remove jest from prod code

afharo reviewed (commented):
I love the purpose of this PR! And I'm happy to approve it as is. 

I'd like to raise an alternative approach that would reduce the need to add a React component and to play catch-up with the existing EUI components:

How about we define new attributes `data-subscription-source` and `data-subscription-feature` and we use `document.querySelector` to find all the elements with those attributes and append the appropriate listeners to report the `subscription__upsell__impression` and `subscription__upsell__click` events?

WDYT?

maximpn reviewed (commented):
@janmonschke thanks for the PR. The changes look nice, it's really a good idea to implement reusable functionality for upsells. I'd like to highlight a few moments before merging it back.

Basically what we want to achieve is having a simple way to track upsell links/buttons impressions and clicks. Shared components like `SubscriptionLink` will do the trick though there are some limitations for example Cmd+Click behaviour which isn't supported by EUI but is nice to have feature. There is an issue example https://github.com/elastic/kibana/issues/99077. Another question if we need to collect analytics for the links leading to another pages.

I see two ways which may have sense here

- from the changes in the current PR it's clear that the whole functionality can be in `useGoToSubscription`
- create a wrapper React component which will react on click event and send statistics. This way the concerns are split so analytics and business logic can be implemented in different components.

To be honest my choice would be investigation of the wrapper component since it perfectly fits into composability pattern and reduces the amount of changes.

janmonschke commented on packages/kbn-subscription-tracking/services.tsx:
```diff
@@ -0,0 +1,100 @@
+  );
+};
+
+const analyticsClientMock: AnalyticsClient = {
+  optIn: () => {},
+  reportEvent: () => {},
+  isEventTypeRegistered: () => false,
+  registerEventType: () => {},
+  registerContextProvider: () => {},
+  removeContextProvider: () => {},
+  registerShipper: () => {},
+  telemetryCounter$: new Subject(),
+  shutdown: () => {},
+};
+
+/**
+ * Mock for the external services provider. Only use in tests!
+ */
+export const MockSubscriptionTrackingProvider: FC = ({ children }) => {
+  return (
+    <SubscriptionTrackingProvider navigateToApp={() => {}} analyticsClient={analyticsClientMock}>
+      {children}
+    </SubscriptionTrackingProvider>
+  );
+};
```
Are there best practices around this documented somewhere? E.g. Should the import be `@kbn/mypackage/mocks`? Or would it be okay to `import { MockThing } from `@kbn/mypackage`?

janmonschke commented:
> How about we define new attributes `data-subscription-source` and `data-subscription-feature` and we use `document.querySelector` to find all the elements with those attributes and append the appropriate listeners to report the `subscription__upsell__impression` and `subscription__upsell__click` events?
> 
> WDYT?

I'm intrigued by this alternative approach that would not need custom React wrappers. Could you go more into detail on how you would implement this? Right now I'm not sure where, when and how exactly we would execute the `document.querySelector` calls 😅  (@afharo)

logeekal commented on packages/kbn-subscription-tracking/types.ts:
```diff
@@ -0,0 +1,47 @@
+
+enum SolutionIdentifier {
+  observability = 'observability',
+  security = 'security',
```
Do you think it makes sense to use `APP_ID` or `APP_UI_ID` here?

janmonschke committed:
typos

janmonschke committed:
remove isEventTypeRegistered

removes isEventTypeRegistered and in turn enables the subscription events globally

PhilippeOberti reviewed (changes requested):
@janmonschke the code and the UI look good, but the change broke our paywall story book story. Let me know if you want any help fixing it!
![Screen Shot 2022-11-02 at 10 29 36 AM](https://user-images.githubusercontent.com/17276605/199532105-929510b5-cb09-453d-8021-b13dc8cbd1d3.png)

afharo commented:
> I'm intrigued by this alternative approach that would not need custom React wrappers. Could you go more into detail on how you would implement this? Right now I'm not sure where, when and how exactly we would execute the `document.querySelector` calls 😅 (@afharo)

@janmonschke, not really sure, to be fair. It was a happy thought... but I realize we don't have a `rerender` event to subscribe to and reevaluate. I think that, for the purpose of this, we can merge this solution and iterate into making it more generic if we see fit

janmonschke commented:
Closed in favor of https://github.com/elastic/kibana/pull/157392
